### PR TITLE
[Extensions] Migrates AnomalyResultAction, EntityResultAction, RCFResultAction

### PR DESCRIPTION
**Pull request by joshpalis**
### Description
Part of https://github.com/opensearch-project/opensearch-sdk-java/issues/383

This PR registers the actions that execute an anomaly detector and facilitate the indexing of anomaly results. Currently, the job configuration, scheduling, execution and result indexing for `SINGLE ENTITY` detectors for `REAL-TIME` analysis is complete. 

Some example logs of the single entity real time analysis job execution.
```
02:09:44.780 [httpclient-dispatch-1] INFO  org.opensearch.ad.transport.AnomalyResultTransportAction - Sending RCF request to ad-extension-1 for model OBMtc4cBmBpoB0ddeoRK_model_rcf_0
02:09:44.780 [httpclient-dispatch-1] INFO  org.opensearch.ad.transport.RCFResultTransportAction - Serve rcf request for OBMtc4cBmBpoB0ddeoRK_model_rcf_0
02:09:44.788 [httpclient-dispatch-1] DEBUG org.opensearch.ad.AnomalyDetectorJobRunner - Update latest realtime task for SINGLE detector OBMtc4cBmBpoB0ddeoRK, total updates: 1444
02:09:44.808 [httpclient-dispatch-5] DEBUG org.opensearch.ad.transport.handler.AnomalyIndexHandler - Succeed in saving OBMtc4cBmBpoB0ddeoRK
02:09:44.809 [httpclient-dispatch-1] INFO  org.opensearch.ad.AnomalyDetectorJobRunner - Released lock for AD job OBMtc4cBmBpoB0ddeoRK
```

Preparational work for `MUTLI ENTITY` detectors for `REAL-TIME` analysis has been done. This PR enables creating a multi-entity detector, scheduling and executing a job. The remaining item is to generate and index the anomaly results for HCAD (the blockers for multi-entity real time analysis is noted in the issue). At this state, when a multi-entity detector job is run, the job execution will end while attempting to index an anomaly result and begin again in the next iteration. 

Note, the hashring has been removed from the `RCFResultTransportAction` and the `AnomalyResultTransportAction`. All requests to run the detector will be executed by the AD extension node.

## Overview of the job execution process for single entity real time analysis
1) Ingest data and create single entity detector for real time analysis
2) Start Detector for real time analysis, schedules AD job
3) `ADJobRunner` receives `runJob`() request from `Job Scheduler`, sends a request to acquire a lock and then first job execution begins and triggers the `AnomalyResultAction`
4) enabled features are processed, verifies data within the detection window is available (this checks if there are any entries in which the `@timestamp` value falls between the detection window)
5) `RCFResultAction` is triggered via the client for extension node
6) `modelManager `is used to retrieve `TRcfResult` for the model ID associated with the detector
7) `checkpointDao` is used to retrieve `TRCFModel` (This returns nothing as checkpoint index is not initialized and there is no model that has been trained)
8) Attempting to process the model will throw `ResourceNotFoundException` since the checkpoint is not initialized and the model is not present
9) Exception failure returns to `rcfActionListener.onFailure()` which handles the prediction failure, and then invokes `coldStartIfNoModel`
10) Exception is then checked if instance of `ResourceNotFoundException` thrown by the model manager
11) If there are no previous exceptions that were thrown due to other cold start attempts, cold start is triggered
12) Within cold start, the model manager then trains a model
13) The trained model is then taken by the checkpoint DAO, determines if the checkpoint exists (if not it initializes it), and then indexes the model checkpoint
14) If no exceptions occurs in cold start, then `rcfActionListener` triggers `AnomalyResultTransportAction listener.onFailure()` to return no model exception (model not ready)
15) The exception is then returned to the `ADJobRunner`, which then indexes exception to `AnomalyResultIndex` (creates it if not present)
16) The lock is then released and the next execution of the job is scheduled
17) The next execution is triggered, steps 3 - 7 repeats, and the model is retrieved from the checkpoint index
18) The model is then used to retrieve the `ThreshHoldedRandomCutForest `result and returned as an `RCFResultResponse`
19) This response is then eventually indexed into the `AnomalyResult` Index and the next job execution is scheduled

## How to run data generation script

1. Modify this [line](https://github.com/opensearch-project/anomaly-detection/blob/a51ce136fa30af747acd6f5bbef8a7326c493d4d/dataGeneration/generate-cosine-data-multi-entity.py#L93) and set the refresh interval to `"1s"` 
2. Follow this [README](https://github.com/opensearch-project/anomaly-detection/blob/feature/extensions/dataGeneration/README.md) to install the necessary requirements
3. Start `AnomalyDetectionExtension` and OpenSearch with Job Scheduler installed (enable feature flag)
4. start data ingestion into the local cluster : `python3 generate-cosine-data-multi-entity.py -ep localhost -i server_log --shards 5 -t 10 --no-security -nh 5 -np 5 --ingestion-frequency 10 --points 30000`  (will take about 30 seconds)

## Detector Configurations Used to test real time analysis
Once the data in ingested, then we can create a detector for the created index `server_log` with the following configurations. The data generation script provides test data with two feature fields ( `cpuTime` and `jvmGcTime` ), and two category fields ( `host` and `process` ) that can be used for HCAD. Once the detector is created, we can start the detector

Single-Entity Detector :
```
{
  "name": "test-detector",
  "description": "Test detector",
  "time_field": "@timestamp",
  "indices": [
    "server_log"
  ],
  "feature_attributes": [
    {
      "feature_name": "test",
      "feature_enabled": true,
      "aggregation_query": {
        "test": {
          "avg": {
            "field": "jvmGcTime"
          }
        }
      }
    }
  ],
  "detection_interval": {
    "period": {
      "interval": 1,
      "unit": "Minutes"
    }
  },
  "window_delay": {
    "period": {
      "interval": 1,
      "unit": "Minutes"
    }
  }
}
```

Multi-Entity Detector
```
{
  "name": "test-detector",
  "description": "Test detector",
  "time_field": "@timestamp",
  "indices": [
    "server_log"
  ],
  "feature_attributes": [
    {
      "feature_name": "test",
      "feature_enabled": true,
      "aggregation_query": {
        "test": {
          "avg": {
            "field": "cpuTime"
          }
        }
      }
    }
  ],
  "detection_interval": {
    "period": {
      "interval": 1,
      "unit": "Minutes"
    }
  },
  "window_delay": {
    "period": {
      "interval": 1,
      "unit": "Minutes"
    }
  },
  "category_field": [
    "host"
  ]
}
```

### Issues Resolved
https://github.com/opensearch-project/opensearch-sdk-java/issues/626

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/transport/ProfileTransportAction.java
Comment (vibrantvarun) [inline src/main/java/org/opensearch/ad/transport/ProfileTransportAction.java:48]: clusterService to sdkClusterService.
client to sdkRestClient.

Please ensure the naming convention on variables.

File: src/main/java/org/opensearch/ad/ml/CheckpointDao.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/ad/ml/CheckpointDao.java:694]: We should check for the index not found message rather directly checking for `OpenSearchStatusException` here?

File: src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:891]: Why are we adding an extra check here?

File: src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:1132]: Same as above for the exception message

File: src/main/java/org/opensearch/ad/transport/RCFResultTransportAction.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/ad/transport/RCFResultTransportAction.java:68]: Where are we using `remoteNodeId` here?

File: src/test/java/org/opensearch/ad/transport/ADResultBulkTransportActionTests.java
Comment (owaiskazi19) [inline src/test/java/org/opensearch/ad/transport/ADResultBulkTransportActionTests.java:11]: Why are we commenting the whole test file for hashring? Can't we just pass null for hashring or use extensioNode instead of making request to all the nodes?

File: src/test/java/org/opensearch/ad/transport/AnomalyResultTests.java
Comment (owaiskazi19) [inline src/test/java/org/opensearch/ad/transport/AnomalyResultTests.java:11]: Same as above

**Review by owaiskazi19 — COMMENTED** [review]
Overall looks good. With few comments

File: src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java
Comment (joshpalis) [reply src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java:626]: Yes, the aggregation returned from the RHLC is of type ParsedDateRange

File: src/main/java/org/opensearch/ad/ml/CheckpointDao.java
Comment (joshpalis) [reply src/main/java/org/opensearch/ad/ml/CheckpointDao.java:694]: sure ill make this change

File: src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java
Comment (joshpalis) [reply src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:891]: Originally the exception returned by the transport service call to the `RCFResultAction` when we attempt to retrieve and process a model from the checkpoint index is a `ResourceNotFoundException` thrown by the `ModelManager` . Since we're now using the sdkClient to execute the request, the returned exception is of type `CompletionException`. Since the `ResourceNotFoundException` is used to trigger cold start, I have opted to convert this exception back to a `ResourceNotFoundException`

File: src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java
Comment (joshpalis) [reply src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:1132]: I can change this to check if the `cause` message contains the string `type=index_not_found_exception` instead

File: src/main/java/org/opensearch/ad/transport/RCFResultTransportAction.java
Comment (joshpalis) [reply src/main/java/org/opensearch/ad/transport/RCFResultTransportAction.java:68]: I kept this to reduce the diff but I can change it

File: src/test/java/org/opensearch/ad/transport/ADResultBulkTransportActionTests.java
Comment (joshpalis) [reply src/test/java/org/opensearch/ad/transport/ADResultBulkTransportActionTests.java:11]: I can handle these changes in the multi-entity results PR. As part of that PR I have modified the `ADResultBulkTransportAction` and removed the indexing pressure object from that class so some of these tests will have to be rewritten

Comment (owaiskazi19) [reply src/test/java/org/opensearch/ad/transport/ADResultBulkTransportActionTests.java:11]: That works

File: src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java
Comment (owaiskazi19) [reply src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:1132]: Let's do that

File: src/test/java/org/opensearch/ad/transport/AnomalyResultTests.java
Comment (joshpalis) [reply src/test/java/org/opensearch/ad/transport/AnomalyResultTests.java:11]: I have commented this out for the hashring since it set up test nodes and registers handlers for these nodes. The hashring is then used to stub calls to return the test nodes. Since we're temporarily rerouting all requests back to the extension node, most of these test classes dont even apply.

Comment (owaiskazi19) [reply src/test/java/org/opensearch/ad/transport/AnomalyResultTests.java:11]: Not blocking this PR for the above but we should think about the tests related to HashRing. We can modify the tests for extensionNode.

**Review by owaiskazi19 — APPROVED** [review]
LGTM! If @dbwiddis can also take a look!

File: build.gradle
Comment (saratvemulapalli) [inline build.gradle:780]: Curious why these classes are not having test coverage?

Comment (joshpalis) [reply build.gradle:780]: The classes regarding ADResultBulkTransportAction have been commented out for this PR temporarily, I will handle them in [this PR](https://github.com/opensearch-project/anomaly-detection/pull/876) to enable HCAD real time analysis, as the ADResultBulkTransportAction is used to index multi entity Anomaly Results. This current PR includes some preparational work for the HCAD workflow

**Review by dbwiddis — APPROVED** [review]

**Pull request merged**
